<!DOCTYPE html>
<html class="writer-html5" lang="en" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.17.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Smoother Manual &mdash; smoother 0.0.1 documentation</title>
      <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
    <link rel="shortcut icon" href="_static/favicon.ico"/>
  <!--[if lt IE 9]>
    <script src="_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
        <script src="_static/jquery.js"></script>
        <script src="_static/underscore.js"></script>
        <script src="_static/doctools.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >
            <a href="index.html" class="icon icon-home"> smoother
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <ul>
<li class="toctree-l1"><a class="reference internal" href="Readme.html">Smoother</a></li>
<li class="toctree-l1"><a class="reference internal" href="Manual.html">Smoother Manual</a></li>
<li class="toctree-l1"><a class="reference internal" href="Preprocess.html">Documentation for the Preprocessing script</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">smoother</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home"></a> &raquo;</li>
      <li>Smoother Manual</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/Manual_.md.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="smoother-manual">
<h1>Smoother Manual<a class="headerlink" href="#smoother-manual" title="Permalink to this headline"></a></h1>
<p>Bulk of the manual</p>
<section id="table-of-contents">
<h2>Table of Contents<a class="headerlink" href="#table-of-contents" title="Permalink to this headline"></a></h2>
</section>
<section id="importing-data">
<h2>Importing Data<a class="headerlink" href="#importing-data" title="Permalink to this headline"></a></h2>
<section id="preprocessing-data-anna">
<h3>Preprocessing Data (<em>Anna</em>)<a class="headerlink" href="#preprocessing-data-anna" title="Permalink to this headline"></a></h3>
<p>generating the chromosome lengths file
annotation file
bed &amp; wig files
input formats
output all alignments
provide some sample files</p>
</section>
<section id="create-index-anna">
<h3>Create Index (<em>Anna</em>)<a class="headerlink" href="#create-index-anna" title="Permalink to this headline"></a></h3>
<p>creating an empty index
adding genome sizes
and annotation file
min bin size parameter</p>
</section>
<section id="adding-replicates-anna">
<h3>Adding Replicates (<em>Anna</em>)<a class="headerlink" href="#adding-replicates-anna" title="Permalink to this headline"></a></h3>
<p>adding hi-c or radicl-seq replicates</p>
</section>
<section id="adding-normalization-tracks-anna">
<h3>Adding Normalization Tracks (<em>Anna</em>)<a class="headerlink" href="#adding-normalization-tracks-anna" title="Permalink to this headline"></a></h3>
<p>adding read or wig normalization tracks</p>
</section>
<section id="adding-grid-seq-normalizations-todo-better-name-markus">
<h3>Adding Grid-seq normalizations (&#64;todo better name) (<em>Markus</em>)<a class="headerlink" href="#adding-grid-seq-normalizations-todo-better-name-markus" title="Permalink to this headline"></a></h3>
<p>creating relevant plots (infliction point)
setting parameters
adding as normalization track</p>
</section>
</section>
<section id="viewing-data">
<h2>Viewing Data<a class="headerlink" href="#viewing-data" title="Permalink to this headline"></a></h2>
<section id="example-experiments-anna">
<h3>Example Experiments (<em>Anna</em>)<a class="headerlink" href="#example-experiments-anna" title="Permalink to this headline"></a></h3>
</section>
<section id="list-of-plots-and-what-they-show-anna">
<h3>List of Plots and what they show (<em>Anna</em>)<a class="headerlink" href="#list-of-plots-and-what-they-show-anna" title="Permalink to this headline"></a></h3>
<p>the heatmap the ratio &amp; coverage plots
mention hovers
also include status text &amp; command line output</p>
</section>
<section id="list-of-buttons-and-their-functionality-both">
<h3>List of Buttons and their functionality (<em>Both</em>)<a class="headerlink" href="#list-of-buttons-and-their-functionality-both" title="Permalink to this headline"></a></h3>
<p>top to bottom list of things
group by category
don’t forget the bokeh buttons</p>
<section id="normalization-markus">
<h4>Normalization (<em>Markus</em>)<a class="headerlink" href="#normalization-markus" title="Permalink to this headline"></a></h4>
<section id="normalize-by">
<h5>Normalize by<a class="headerlink" href="#normalize-by" title="Permalink to this headline"></a></h5>
<p>Smoother can normalize data in various ways. Some of them are suited for symmetric data (e.g. Hi-C), some for asymmetric data (e.g. RADICL-seq).
For some normalizations, views or different regions or with different bin sizes are not comparable.
Note that for these normalizations, even zooming or moving around may skew the displayed values.
It is recommended to use these normalizations merely as an exploratory tool but never to compare e.g. two screenshots of different regions.</p>
<table class="docutils align-default">
<colgroup>
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
<col style="width: 17%" />
</colgroup>
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Picture</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>For Symm.</p></th>
<th class="head"><p>For Asymm.</p></th>
<th class="head"><p>Always Comparable</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Largest Rendered Bin</p></td>
<td></td>
<td><p>Divide the number of interactions in each bin the number of interactions in the largest rendered bin. This keeps differences between bins nicely visible by always ensuring that the colorscale is used fully.</p></td>
<td><p>✔️</p></td>
<td><p>✔️</p></td>
<td></td>
</tr>
<tr class="row-odd"><td><p>Number of Reads</p></td>
<td></td>
<td><p>Divide the number of interactions in each bin by the total number of reads in the dataset.</p></td>
<td><p>✔️</p></td>
<td><p>✔️</p></td>
<td><p>✔️</p></td>
</tr>
<tr class="row-even"><td><p>Column Sum</p></td>
<td></td>
<td><p>Divide the number of interactions in each bin by the number of interactions in the complete column that bin belongs to.</p></td>
<td></td>
<td><p>✔️</p></td>
<td><p>✔️</p></td>
</tr>
<tr class="row-odd"><td><p>Row Sum</p></td>
<td></td>
<td><p>Divide the number of interactions in each bin by the number of interactions in the complete row that bin belongs to.</p></td>
<td></td>
<td><p>✔️</p></td>
<td><p>✔️</p></td>
</tr>
<tr class="row-even"><td><p>Coverage of Normalization Reads (Absolute)</p></td>
<td></td>
<td><p>Divide the number of interactions in each bin by the coverage of the normalization datasets. With the ‘Normalization Rows’ and ‘Normalization Columns’ pickers, you can decide what datasets should be used for the columns and rows.</p></td>
<td><p>✔️</p></td>
<td><p>✔️</p></td>
<td><p>✔️</p></td>
</tr>
<tr class="row-odd"><td><p>Coverage of Normalization Reads (Scaled)</p></td>
<td></td>
<td><p>Same as the absolute version, but makes sure the complete colorscale is used.</p></td>
<td><p>✔️</p></td>
<td><p>✔️</p></td>
<td></td>
</tr>
<tr class="row-even"><td><p>Binominal Test</p></td>
<td></td>
<td><p>Use a binominal test to determine weather each bin is statistically significant, in it’s row. This strategy was created by Bonetti et al. for RADICL-seq RNA-DNA interaction data [3]. The acceptance p-value can be modified with a slider. <a href="#id1"><span class="problematic" id="id2">*</span></a>For views that do not show the entire genome, we use an approximation for the p-value adjustment that is performed after the binominal test.</p></td>
<td></td>
<td><p>✔️</p></td>
<td><p>✔️*</p></td>
</tr>
<tr class="row-odd"><td><p>Iterative Correction</p></td>
<td></td>
<td><p>Use the iterative correction approach developed for Hi-C data by Imakaev et al. [4].</p></td>
<td><p>✔️</p></td>
<td></td>
<td><p>✔️</p></td>
</tr>
</tbody>
</table>
<p>Since smoother is capable of normalizing data on the fly, you can zoom in to a region of interest and then dynamically change the normalization strategy.</p>
<p>Groups A and B are normalized individually and then combined according to the ‘Between Group’ setting (see Replicates section).
The ‘In Group’ operation (see Replicates section) is applied before the Normalization.</p>
</section>
<section id="mapping-quality-bounds">
<h5>Mapping Quality Bounds<a class="headerlink" href="#mapping-quality-bounds" title="Permalink to this headline"></a></h5>
<p>&#64;todo move to Filters</p>
</section>
<section id="color-scale-begin-color-scale-log-base">
<h5>Color Scale Begin &amp; Color Scale Log Base<a class="headerlink" href="#color-scale-begin-color-scale-log-base" title="Permalink to this headline"></a></h5>
<p>After normalization, the values in each bin are between zero and one.
Before displaying these values in the heatmap, we apply a logarithmic transformation.
Like with logarithmic scales, this keeps all values in the same order but this makes a difference between two small numbers appear bigger than the same difference between two bigger numbers.
You can fiddle with the strength of this effect using the ‘Color Scale Log Base’ slider.
In brief, higher values for the log base increase the difference between small numbers but make larger number look more similar.
Lower values do the opposite.
Setting log base to zero displays the numbers without any transformation.</p>
<p>Additionally, the ‘color scale begin’ slider can be used to cut off the bottom of the color scale, i.e. to only show bins with more than x interactions.
This option subtracts from the raw number of interactions; it is applied before normalization and application of the log scale.</p>
<p>In detail, the function we use is:</p>
<img src="../static/log_scale.png"/><p>, where <em>x</em> and <em>y</em> are the un-normalized and normalized interaction frequencies while <em>a</em> is the log base parameter.
With varying values for <em>a</em>, this function always satisfies <em>f(0)=0</em> and <em>f(1)=1</em>.
for <em>a=0</em> the function is undefined but approaches the 45-degree diagonal, so we hardcode <em>f(x)=x</em> for <em>a=0</em>.</p>
</section>
</section>
<section id="quick-config-buttons">
<h4>Quick Config Buttons<a class="headerlink" href="#quick-config-buttons" title="Permalink to this headline"></a></h4>
<p>what do they change</p>
</section>
</section>
</section>
<section id="setting-up-a-webserver-markus">
<h2>Setting up a Webserver (<em>Markus</em>)<a class="headerlink" href="#setting-up-a-webserver-markus" title="Permalink to this headline"></a></h2>
<p>smoother can be deployed as a webserver so this should be described</p>
</section>
<section id="using-slurm-markus">
<h2>Using Slurm (<em>Markus</em>)<a class="headerlink" href="#using-slurm-markus" title="Permalink to this headline"></a></h2>
<p>smoother can run within slurm</p>
</section>
<section id="installing-via-github-markus">
<h2>Installing via GitHub (<em>Markus</em>)<a class="headerlink" href="#installing-via-github-markus" title="Permalink to this headline"></a></h2>
<p>getting the current development version</p>
</section>
<section id="the-datastructure-markus">
<h2>The Datastructure (<em>Markus</em>)<a class="headerlink" href="#the-datastructure-markus" title="Permalink to this headline"></a></h2>
<p>link to datastructure github</p>
<section id="brief-algorithmic-description">
<h3>Brief Algorithmic Description<a class="headerlink" href="#brief-algorithmic-description" title="Permalink to this headline"></a></h3>
</section>
<section id="fileformat-specification">
<h3>Fileformat specification<a class="headerlink" href="#fileformat-specification" title="Permalink to this headline"></a></h3>
<p>what files
how are they encoded (link to libSps)</p>
</section>
<section id="implementation-details">
<h3>Implementation Details<a class="headerlink" href="#implementation-details" title="Permalink to this headline"></a></h3>
<p>some runtime details</p>
</section>
</section>
<section id="citing-smoother">
<h2>Citing Smoother<a class="headerlink" href="#citing-smoother" title="Permalink to this headline"></a></h2>
</section>
<section id="references">
<h2>References<a class="headerlink" href="#references" title="Permalink to this headline"></a></h2>
<p>[1] Shekelyan, M., Dignös, A. &amp; Gamper, J. Sparse prefix sums: Constant-time range sum queries over sparse multidimensional data cubes. Information Systems 82, 136–147 (2019).</p>
<p>[2] Schmidt et al. &#64;todo</p>
<p>[3] Bonetti, A. et al. RADICL-seq identifies general and cell type–specific principles of genome-wide RNA-chromatin interactions. Nat Commun 11, 1018 (2020).</p>
<p>[4] Imakaev, M. et al. Iterative correction of Hi-C data reveals hallmarks of chromosome organization. Nat Methods 9, 999–1003 (2012).</p>
</section>
<section id="thanks-to">
<h2>Thanks to…<a class="headerlink" href="#thanks-to" title="Permalink to this headline"></a></h2>
<ul class="simple">
<li><p>Bokeh</p></li>
<li><p>Stxxl</p></li>
<li><p>Sparse Prefix Sums paper</p></li>
<li><p>Papers with the normalizations</p></li>
<li><p>Dataset origin</p></li>
</ul>
</section>
</section>


           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2022, Markus Schmidt.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>